assembly/src/ast.rs

```diff
@@ -0,0 +1,141 @@
+// AST
+// ================================================================================================
+
+#[derive(PartialEq, Eq)]
+pub enum PushMode {
+    Adv(u64),
+    EnvLocaddr(u64),
+    EnvSdepth,    
+    Local(u64),
+    Mem(u64)
+}
+
+// This holds the list of operations supported in a Miden program.
+pub enum Instruction {
+    Assert,
+    AssertEq,
+    Add,
+    AddImm(u64),
+    Sub,
+    SubImm(u64),
+    Mul,
+    MulImm(u64),
+    Div,
+    DivImm(u64),
+    Neg,
+    Inv,
+    Pow2(bool),
+    Not,
+    And,
+    Or,
+    Xor,
+    Eq,
+    EqImm(u64),
+    Neq,
+    NeqImm(u64),
+    Eqw,
+    Lt,
+    Lte,
+    Gt,
+    Gte,
+
+    // ----- u32 manipulation ---------------------------------------------------------------
+
+    U32Test,
+    U32TestW,
+    U32Assert,
+    U32AssertW,
+    U32Split,
+    U32Cast,
+    U32CheckedAdd,
+    U32WrappingAdd,
+    U32OverflowingAdd,   
+    U32OverflowingAdd3,
+    U32CheckedSub,
+    U32WrappingSub,
+    U32OverflowingSub,
+    U32CheckedMul,
+    U32WrappingMul,
+    U32OverflowingMul,
+    U32OverflowingMadd,
+    U32Div(bool),
+    U32Mod(bool),
+    U32DivMod(bool),
+    U32CheckedAnd,
+    U32CheckedOr,
+    U32CheckedXor,
+    U32CheckedNot,
+    U32Shr(bool),
+    U32Shl(bool),
+    U32Rotr(bool),
+    U32Rotl(bool),
+    U32CheckedEq,
+    U32CheckedNeq,
+    U32Lt(bool),
+    U32Lte(bool),
+    U32Gt(bool),
+    U32Gte(bool),
+    U32Min(bool),
+    U32Max(bool),
+
+    // ----- stack manipulation ---------------------------------------------------------------
+    Drop,
+    DropW,
+    PadW,
+    Dup,
+    DupW,
+    SwapW,
+    SwapDW,
+    MovUp,
+    MovUpW,
+    MovDn,
+    MovDnW,
+    CSwap,
+    CSwapW,
+    CDrop,
+    CDropW,
+
+    // ----- input / output operations --------------------------------------------------------
+    Push(PushMode),
+    PushWLocal(Option<u64>),
+    PushWMem(Option<u64>),
+    PopLocal(u64),
+    PopMem(Option<u64>),
+    PopWMem(Option<u64>),
+    PopWLocal(u64),
+    LoadWAdv,
+    LoadWMem(Option<u64>),
+    LoadWLocal(u64),
+    StoreWMem(Option<u64>),
+    StoreWLocal(u64),
+    AdvU64Div,
+
+    // ----- cryptographic operations ---------------------------------------------------------
+    RPHash,
+    RPPerm,
+    MTreeGet,
+    MTreeSet,
+    MTreeCWM
+}        
+
+// A node in a AST that can represent a instruction or a control flow.
+pub enum Node {
+    Instruction(Instruction),
+    IfElse(Vec<Node>, Vec<Node>),
+    Repeat(Vec<Node>),
+    Call,
+    While(Vec<Node>),
+}
+
+// Procedure holds information about a defnied procedure in a Miden program.
+pub struct Procedure {
+
+}
+
+// Represents a parsed program AST.
+// This AST can then be furthur processed to generate MAST.
+pub struct Program {
+    pub modules: Vec<String>,
+    pub procedures: Vec<Procedure>,
```

Comment:
This should probably be a map to enable lookups.

Reply:
fixed